### PR TITLE
Sync Permits

### LINKED ISSUE #5776: Create the sync permit logic for the new sync protocol
Implement a class that will provide sync permits, with the maximum configured at construction. This will be used by each individual peer negotiator thread when sync is implemented as a `Protocol`

### PR DESCRIPTION
**Pull request by poulok**
**Description**:
Add the `SyncPermit` class in preparation for #5702 

**Related issue(s)**:

Fixes #5776

**Notes for reviewer**:
This code is not currently used except for in unit tests.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        return syncPermits.availablePermits();
+    }
+
+    /**
+     * Attempts to acquire a sync permit. This method returns immediately if no permit is not available.
+     *
+     * @return an autocloseable instance that tells the caller if the permit has been acquired and will automatically
+     * release the permit when used in a try-with-resources block
+     */
+    public MaybeLocked tryAcquire() {
```
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java:55]:
```suggestion
    public @NonNull MaybeLocked tryAcquire() {
```

**Review by cody-littley — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java (hunk 1 above)
Comment (hendrikebbers) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java:55]: +1

**Review by hendrikebbers — DISMISSED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java (hunk 1 above)
Comment (poulok) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java]: done

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * @return the current number of available sync permits
+     */
+    public int getNumAvailable() {
+        return syncPermits.availablePermits();
+    }
+
+    /**
+     * Attempts to acquire a sync permit. This method returns immediately if no permit is not available.
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java:51]: it returns immediately either way, right?

**Review by cody-littley — APPROVED** [review]

**Review by hendrikebbers — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java (hunk 2 above)
Comment (poulok) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/SyncPermit.java:51]: yes

**Pull request merged**